src/effects/deeplearning/ModelCard.cpp

```diff
@@ -0,0 +1,340 @@
+{
+
+   void validateExists(const std::string &key, DocHolder doc)
+   {
+      if(!doc->IsObject())
+         throw InvalidModelCardDocument("provided document is not an object", doc);
+      if(!doc->HasMember(key.c_str()))
+      {
+         wxString msg = wxString("JSON document missing field: %s").Format(wxString(key));
+         throw InvalidModelCardDocument(msg.ToStdString(), doc);
+      }
+   }
+
+   void throwTypeError(const std::string &key, const char *type, DocHolder doc)
+   {
+      wxString msg = wxString("field: %s is not of type: %s").Format(wxString(key), wxString(type));
+      throw InvalidModelCardDocument(msg.ToStdString(), doc);
+   }
+
+   std::string tryGetString(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsString())
+         throwTypeError(key, "string", doc);
+
+      return (*doc)[key.c_str()].GetString();
+   }
+
+   int tryGetInt(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsInt())
+         throwTypeError(key, "int", doc);
+      
+      return (*doc)[key.c_str()].GetInt();
+   }
+
+   bool tryGetBool(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsBool())
+         throwTypeError(key, "bool", doc);
+      
+      return (*doc)[key.c_str()].GetBool();
+   }
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsArray())
+         throwTypeError(key, "array", doc);
+      
+      std::vector<std::string> labels;
+      for (auto itr = (*doc)[key.c_str()].Begin(); itr != (*doc)[key.c_str()].End();  ++itr)
+         labels.emplace_back(itr->GetString());
+
+      return labels;
+   }
+
+   uint64_t tryGetUint64(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsUint64())
+         throwTypeError(key, "uint64", doc);
+      
+      return (*doc)[key.c_str()].GetUint64();
+   }
+
+}
+
+namespace parsers 
+{
+   DocHolder ParseString(const std::string &data)
+   {
+      DocHolder d = std::make_shared<rapidjson::Document>();
+      // parse the data
+      d->Parse(data.c_str());
+      if (d->Parse(data.c_str()).HasParseError()) 
+      {
+         std::string msg = "error parsing JSON from string: \n";
+         msg += rapidjson::GetParseError_En(d->GetParseError());
+         msg += "\n document: \n\t" + std::string(data.c_str()) + "\n";
+         throw InvalidModelCardDocument(msg.c_str(), d);
+      }
+      
+      return d;
+   }
+
+   DocHolder ParseFile(const std::string &path)
+   {
+      std::shared_ptr<wxString> docStr = std::make_shared<wxString>();
+      wxFile file = wxFile(path);
+
+      if(!file.ReadAll(docStr.get()))
+         throw InvalidModelCardDocument("could not read file", nullptr);
+
+      return ParseString(docStr->ToStdString());
```

Comment:
I would prefer explicit encoding here. lib-strings can help with it :-)

Reply:
Hmm, I'm not entirely sure what you mean. Would you mind clarifying? Thanks!

Reply:
`audacity::ToUTF8` from `CodeConversions.h`